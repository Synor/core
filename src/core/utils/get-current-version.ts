import { sortVersions } from 'utils/sort-versions'

type SynorMigrationRecord = import('../migration').SynorMigrationRecord

type Version = SynorMigrationRecord['version']

export function getCurrentVersion(history: SynorMigrationRecord[]): string {
  const appliedVersionMap = history.reduce<Record<Version, true>>(
    (applied, { version, type }): Record<Version, true> => {
      if (type === 'DO') {
        applied[version] = true
      }

      if (type === 'UNDO') {
        delete applied[version]
      }

      return applied
    },
    {}
  )

  const appliedVersions = sortVersions(Object.keys(appliedVersionMap))

  const currentVersion = appliedVersions[appliedVersions.length - 1]

  return currentVersion
}
