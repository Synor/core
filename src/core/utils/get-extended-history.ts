type MigrationRecord = import('../migration').MigrationRecord
type ExtendedMigrationRecord = import('../migration').ExtendedMigrationRecord

type ID = MigrationRecord['id']
type Version = MigrationRecord['version']

export function getExtendedHistory(
  baseVersion: Version,
  history: MigrationRecord[]
): ExtendedMigrationRecord[] {
  const extraInfoById: Record<
    ID,
    Omit<ExtendedMigrationRecord, keyof MigrationRecord>
  > = {}

  const lastRecordIdByVersion: Record<Version, ID> = {}

  for (const { id, version } of history) {
    extraInfoById[id] = {
      state: 'applied',
      revertedBy: null
    }

    if (version === baseVersion) {
      continue
    }

    const prevRecordId = lastRecordIdByVersion[version]

    if (prevRecordId) {
      extraInfoById[prevRecordId].state = 'reverted'
      extraInfoById[prevRecordId].revertedBy = id
    }

    lastRecordIdByVersion[version] = id
  }

  const extendedHistory: ExtendedMigrationRecord[] = history.map(
    ({ id }, index) => ({ ...history[index], ...extraInfoById[id] })
  )

  return extendedHistory
}
