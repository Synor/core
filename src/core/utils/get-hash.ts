import crypto from 'crypto'

export function getHash(content: string): string {
  const sanitizedContent = content.replace(/[\r\n]+/gm, '\n').trim()
  return crypto
    .createHash('sha256')
    .update(sanitizedContent, 'utf8')
    .digest('hex')
}
