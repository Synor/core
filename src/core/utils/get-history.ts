type DatabaseEngine = import('../database').DatabaseEngine
type MigrationRecord = import('../migration').MigrationRecord
type MigrationHistory = import('../migration').MigrationHistory

type MigrationVersion = MigrationRecord['version']
type RecordID = MigrationRecord['id']

export async function getHistory(
  database: DatabaseEngine,
  baseVersion: MigrationVersion,
  recordStartId: RecordID
): Promise<MigrationHistory> {
  const records = await database.records(recordStartId)

  const lastIndexByVersion: Record<MigrationVersion, number> = {}

  const history: MigrationHistory = records.map(record => ({
    ...record,
    state: 'applied',
    revertedBy: null
  }))

  history.forEach(({ id, version }, index) => {
    if (version === baseVersion) {
      return
    }

    const lastIndex = lastIndexByVersion[version]

    if (lastIndex) {
      history[lastIndex].state = 'reverted'
      history[lastIndex].revertedBy = id
    }

    lastIndexByVersion[version] = index
  })

  return history
}
