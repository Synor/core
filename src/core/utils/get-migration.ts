import { SynorMigration } from '../migration'

type MigrationSource = import('../migration').MigrationSource
type MigrationType = import('../migration').MigrationType
type MigrationVersion = import('../migration').MigrationVersion
type SourceEngine = import('../source').SourceEngine

export async function getMigration(
  source: SourceEngine,
  version: MigrationVersion,
  type: MigrationType
): Promise<MigrationSource | null> {
  const info = await source.get(version, type)

  if (!info) {
    return null
  }

  const content = await source.read(info)

  const migration = SynorMigration(info, content)

  return migration
}
