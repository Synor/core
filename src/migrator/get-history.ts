type DatabaseEngine = import('../database').DatabaseEngine
type MigrationHistory = import('../migration').MigrationHistory
type MigrationRecord = import('../migration').MigrationRecord

type RecordID = MigrationRecord['id']

export async function getHistory(
  database: DatabaseEngine,
  baseVersion: string,
  recordStartId: RecordID
): Promise<MigrationHistory> {
  const records = await database.records(recordStartId)

  const lastIndexByVersion: Record<string, number> = {}

  const history: MigrationHistory = records.map(record => ({
    ...record,
    state: 'applied',
    revertedBy: null
  }))

  history.forEach(({ id, version }, index) => {
    if (version === baseVersion) {
      return
    }

    const lastIndex = lastIndexByVersion[version]

    if (lastIndex) {
      history[lastIndex].state = 'reverted'
      history[lastIndex].revertedBy = id
    }

    lastIndexByVersion[version] = index
  })

  return history
}
