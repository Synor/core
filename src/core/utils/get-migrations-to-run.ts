import { getMigration } from './get-migration'

type MigrationSource = import('../migration').MigrationSource
type SourceEngine = import('../source').SourceEngine

type Version = MigrationSource['version']
type Type = MigrationSource['type']

export async function getMigrationsToRun(
  source: SourceEngine,
  baseVersion: Version,
  fromVersion: Version,
  toVersion: Version
): Promise<MigrationSource[]> {
  const migrations: MigrationSource[] = []

  let type: Type

  if (fromVersion < toVersion) {
    type = 'DO'
  } else if (fromVersion > toVersion) {
    type = 'UNDO'
  } else {
    return migrations
  }

  if (fromVersion !== baseVersion) {
    const fromVersionExist = await source.get(fromVersion, type)
    if (!fromVersionExist) {
      throw new Error(`NOT_FOUND: Migration(${fromVersion}.${type})`)
    }
  }

  if (toVersion !== baseVersion) {
    const toVersionExist = await source.get(toVersion, type)
    if (!toVersionExist) {
      throw new Error(`NOT_FOUND: Migration(${toVersion}.${type})`)
    }
  }

  if (type === 'DO') {
    let nextVersion =
      fromVersion === baseVersion
        ? await source.first()
        : await source.next(fromVersion)

    while (nextVersion) {
      const migration = await getMigration(source, nextVersion, type)

      if (!migration) {
        nextVersion = null
        break
      }

      migrations.push(migration)

      if (migration.version === toVersion) {
        nextVersion = null
        break
      }

      nextVersion = await source.next(migration.version)
    }
  }

  if (type === 'UNDO') {
    let prevVersion = fromVersion === baseVersion ? null : fromVersion

    while (prevVersion) {
      const migration = await getMigration(source, prevVersion, type)

      if (!migration) {
        prevVersion = null
        break
      }

      if (migration.version === toVersion) {
        prevVersion = null
        break
      }

      migrations.push(migration)

      prevVersion = await source.prev(migration.version)
    }
  }

  return migrations
}
