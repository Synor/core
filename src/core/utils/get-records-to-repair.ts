import { getMigration } from './get-migration'

type MigrationHistory = import('../migration').MigrationHistory
type SourceEngine = import('../source').SourceEngine

type MismatchedRecord = Pick<MigrationHistory[number], 'id' | 'hash'>

type MigrationVersion = MigrationHistory[number]['version']

export async function getRecordsToRepair(
  source: SourceEngine,
  baseVersion: MigrationVersion,
  history: MigrationHistory
): Promise<MismatchedRecord[]> {
  const appliedRecords = history.filter(
    ({ state, dirty }) => state === 'applied' && !dirty
  )

  const mismatchedRecords: MismatchedRecord[] = []

  for (const { id, version, type, hash } of appliedRecords) {
    if (version === baseVersion) {
      continue
    }

    const migration = await getMigration(source, version, type)

    if (!migration) {
      throw new Error(`NOT_FOUND: Migration(${version}.${type})`)
    }

    if (migration.hash !== hash) {
      mismatchedRecords.push({ id, hash: migration.hash })
    }
  }

  return mismatchedRecords
}
