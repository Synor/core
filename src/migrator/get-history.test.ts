import { getHistory } from './get-history'
import { getMigrationRecord } from './index.utils.test'

const database = {
  records: jest.fn()
}

describe('migrator:getHistory', () => {
  test('works as expected', async () => {
    database.records.mockImplementationOnce(() =>
      Promise.resolve([
        getMigrationRecord({ id: 1, version: '01', type: 'do' }),
        getMigrationRecord({ id: 2, version: '02', type: 'do' }),
        getMigrationRecord({ id: 3, version: '03', type: 'do' }),
        getMigrationRecord({ id: 4, version: '03', type: 'undo' }),
        getMigrationRecord({ id: 5, version: '02', type: 'undo' })
      ])
    )

    await expect(
      getHistory(database as any, '01', 1)
    ).resolves.toMatchSnapshot()
  })
})
