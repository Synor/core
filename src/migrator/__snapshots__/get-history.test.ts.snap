// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrator:getHistory works as expected 1`] = `
Array [
  Object {
    "appliedAt": 2020-01-01T00:00:00.000Z,
    "appliedBy": "Jest",
    "dirty": false,
    "executionTime": 0,
    "hash": "",
    "id": 1,
    "revertedBy": null,
    "state": "applied",
    "title": "",
    "type": "do",
    "version": "01",
  },
  Object {
    "appliedAt": 2020-01-01T00:00:00.000Z,
    "appliedBy": "Jest",
    "dirty": false,
    "executionTime": 0,
    "hash": "",
    "id": 2,
    "revertedBy": 5,
    "state": "reverted",
    "title": "",
    "type": "do",
    "version": "02",
  },
  Object {
    "appliedAt": 2020-01-01T00:00:00.000Z,
    "appliedBy": "Jest",
    "dirty": false,
    "executionTime": 0,
    "hash": "",
    "id": 3,
    "revertedBy": 4,
    "state": "reverted",
    "title": "",
    "type": "do",
    "version": "03",
  },
  Object {
    "appliedAt": 2020-01-01T00:00:00.000Z,
    "appliedBy": "Jest",
    "dirty": false,
    "executionTime": 0,
    "hash": "",
    "id": 4,
    "revertedBy": null,
    "state": "applied",
    "title": "",
    "type": "undo",
    "version": "03",
  },
  Object {
    "appliedAt": 2020-01-01T00:00:00.000Z,
    "appliedBy": "Jest",
    "dirty": false,
    "executionTime": 0,
    "hash": "",
    "id": 5,
    "revertedBy": null,
    "state": "applied",
    "title": "",
    "type": "undo",
    "version": "02",
  },
]
`;
